# Italy-Finland climate round trip with Python

This notebook demonstrates the use of Python for geospatial analysis to explore climate indexes on Italy and Finland. We'll work with climate data from the Copernicus Climate Data Store and ECMWF (European Centre for Medium-Range Weather Forecasts) data store (CDs and XDS).


### Setting up our environment

First, let's install and import all the necessary libraries:

In [ ]:
# Install a pip package in the current Jupyter kernel
import sys
!{sys.executable} -m pip install numpy pandas xarray netCDF4 matplotlib geopandas cdsapi dask cartopy plotly scipy

In [126]:
# Standard library imports
import os

# Data acquisition
import cdsapi

# Data manipulation
import pandas as pd
import numpy as np
import xarray as xr

# Geospatial libraries
import geopandas as gpd

# Plotting libraries
import matplotlib.pyplot as plt

### What is a NetCDF file?

The Copernicus Climate Data Store provides climate data in NetCDF format (Network Common Data Form – .nc extension). It is a data format specifically designed for storing multidimensional scientific data. These files are particularly common in climate science, meteorology, oceanography, and other geosciences. 

The Copernicus data we are using contains gridded temperature observations across Europe, it represents temperature values across different geographical coordinates (latitude, longitude) and time periods.

Key features of NetCDF (.nc) files include:

- Self-describing format - The file contains metadata about the variables, dimensions, units, and other attributes, making it self-documenting.
- Machine-independent - NetCDF files can be accessed across different computing platforms.
- Direct access - You can access a subset of data without reading the entire file.
- Supports arrays - Designed to efficiently store and retrieve multidimensional arrays of data.
- Append-able - New data can be added to an existing NetCDF file without rewriting it.

## Load and explore Italy and Finland mean temperature data

### Retrieve data from Copernicus

The data we will use is a daily gridded meteorological data for Europe from 1950 to present derived from in-situ observations. 
https://cds.climate.copernicus.eu/datasets/insitu-gridded-observations-europe

Gridded data means that the data is arranged in a grid-like pattern, typically consisting of equally spaced points or cells across a geographic area.

We can download data from the Copernicus Climate Data Store with the Climate Data Store (CDS) Application Program Interface (API) https://cds.climate.copernicus.eu/how-to-api. 
To use the API, you should register to the service and store your personal access token to the file $HOME/.cdsapirc (in your Unix/Linux environment). 

To make things easier and faster, we provide ready NetCDF files, that you can download here https://tinyurl.com/pycon-climatetrip.

In [127]:
# Retrieve data from Copernicus
def get_copernicus_data(output_file='tg_ens_mean_0.25deg_reg_v30.0e.nc'):

    if os.path.exists(output_file):
        print(f"File {output_file} already exists. Using cached version.")
        return output_file
        
    dataset = "insitu-gridded-observations-europe"
    request = {
        "product_type": "ensemble_mean",
        "variable": ["mean_temperature"],
        "grid_resolution": "0_25deg",
        "period": "full_period",
        "version": ["30_0e"]
    }
    
    print("Downloading data from Copernicus...")
    client = cdsapi.Client()
    client.retrieve(dataset, request).download()
    print(f"Data downloaded successfully to {output_file}")
    return output_file

In [ ]:
get_copernicus_data()

### Read and load the data

In [ ]:
# Xarray to read the NetCDF file and load it as a labeled Dataset
ds = xr.open_dataset('./tg_ens_mean_0.25deg_reg_v30.0e.nc')

print("Dataset Overview:")
display(ds)

print(f"\nTime period: {ds.time[0].values} to {ds.time[-1].values}")

print("Dataset spatial coverage:")
print(f"Latitude:  {ds.latitude.min().values:.2f}°N to {ds.latitude.max().values:.2f}°N (resolution: 0.25°)")
print(f"Longitude: {ds.longitude.min().values:.2f}°E to {ds.longitude.max().values:.2f}°E (resolution: 0.25°)")
print(f"Grid size: {len(ds.latitude)} x {len(ds.longitude)} points")

Dataset Overview:


<xarray.Dataset> Size: 10GB
Dimensions:    (latitude: 201, longitude: 464, time: 27210)
Coordinates:
  * latitude   (latitude) float64 2kB 25.38 25.62 25.88 ... 74.88 75.12 75.38
  * longitude  (longitude) float64 4kB -40.38 -40.12 -39.88 ... 75.12 75.38
  * time       (time) datetime64[ns] 218kB 1950-01-01 1950-01-02 ... 2024-06-30
Data variables:
    tg         (time, latitude, longitude) float32 10GB ...
Attributes:
    E-OBS_version:  30.0e
    Conventions:    CF-1.4
    References:     http://surfobs.climate.copernicus.eu/dataaccess/access_eo...
    history:        Fri Aug 30 12:35:42 2024: ncks --no-abc -d time,0,27209 /...
    NCO:            netCDF Operators version 5.1.8 (Homepage = http://nco.sf....


Time period: 1950-01-01T00:00:00.000000000 to 2024-06-30T00:00:00.000000000
tg variable's attributes: {'units': 'Celsius', 'long_name': 'mean temperature', 'standard_name': 'air_temperature'}
Dataset spatial coverage:
Latitude:  25.38°N to 75.38°N (resolution: 0.25°)
Longitude: -40.38°E to 75.38°E (resolution: 0.25°)
Grid size: 201 x 464 points


#### Coordinate reference system and projections

A coordinate reference system (CRS) defines how spatial data corresponds to real locations on Earth.

The CRS is important because the geometric shapes in a GeoSeries or GeoDataFrame object are simply a collection of coordinates in an arbitrary space. A CRS tells Python how those coordinates relate to places on the Earth.

The most common global datum is WGS84 (World Geodetic System 1984), which is used by GPS and serves as the standard for many global datasets, including most Copernicus products.

Map projection is a technique used to display the spherical shape of the Earth on a flat surface. It’s a conversion of the 3D shape of Earth to 2D form. There are different techniques used for map projection.

### Prepare geographic boundaries for Italy and Finland

Using Geopandas, we'll load the country boundaries to focus our analysis on these two European countries.

In [ ]:
# Downloads world boundary data from Natural Earth (a public domain geographic dataset) - The 110m in the URL indicates this is a low-resolution version suitable for overview maps
# https://www.naturalearthdata.com/downloads/110m-cultural-vectors/
url = "https://naturalearth.s3.amazonaws.com/110m_cultural/ne_110m_admin_0_countries.zip"
# Uses geopandas to read the zipped shapefile directly from the URL
world = gpd.read_file(url)

# Filters the world dataset to keep only Italy and Finland
countries = world[world['NAME'].isin(['Italy', 'Finland'])]
# Transform all geometries to the WGS84 coordinate reference system. IN this case the geometries are already in WGS84, so the step is unecessary
# countries = countries.to_crs(epsg=4326)

# Create a simple plot to check the boundaries of the countries
countries.plot()
plt.show()

# Separates Italy and Finland into individual GeoDataFrames
italy = countries[countries['NAME'] == 'Italy']
finland = countries[countries['NAME'] == 'Finland']

# Gets the bounding box coordinates for each country using total_bounds
italy_bounds = italy.total_bounds
finland_bounds = finland.total_bounds

# The bounds are returned as (minx, miny, maxx, maxy)
print(f"Italy bounds: {italy_bounds}")
print(f"Finland bounds: {finland_bounds}")

### Extract climate data for our regions of interest

Now we'll use the geographical boundaries to extract mean temperature for Italy and Finland. We select from the dataset the data for Italy and Finland, by selection only the latitude and longitude values between their bounding boxes calculated previously.

In [ ]:
# Function to extract climate data for a specific geographical region
def extract_country_data(dataset, bounds):
    # Unpack the bounding box coordinates
    minx, miny, maxx, maxy = bounds
    
    # Use xarray's sel method to slice the data:
    # - latitude: select values between miny and maxy
    # - longitude: select values between minx and maxx
    return dataset.sel(latitude=slice(miny, maxy), longitude=slice(minx, maxx))

# Extract data for Italy and Finland using their respective bounding boxes
italy_data = extract_country_data(ds, italy_bounds)
finland_data = extract_country_data(ds, finland_bounds)

# Print the shape of the mean temperature extracted data for Italy
print(f"Italy's tg shape: {italy_data.tg.shape}")
# Check if the extracted data contains any valid (non-NaN) values
print(f"Has valid data: {np.any(~np.isnan(italy_data.tg)).values}")
print(f"Contains NaN: {np.isnan(italy_data.tg).all().values}")

# Print the shape of the mean temperature extracted data for Finland
print(f"Finland's tg shape: {finland_data.tg.shape}")
# Check if the extracted data contains any valid (non-NaN) values
print(f"Has valid data: {np.any(~np.isnan(finland_data.tg)).values}")
print(f"Contains NaN: {np.isnan(finland_data.tg).all().values}")

#### Data description


Italy's tg variable will be a multidimensional 3D array with shape (27210, 42, 47):

First dimension is time (27210 days)

Second dimension is latitude (42 grid points)

Third dimension is longitude (47 grid points)

Each value in this array represents the mean temperature at a specific location and time in Italy. 

In [ ]:
# Check if climate data and boundary data use the same coordinate system and if the coordinates are the same
print(f"Climate data coordinates: {italy_data.longitude.min().values:.2f}  {italy_data.latitude.min().values:.2f} {italy_data.longitude.max().values:.2f}  {italy_data.latitude.max().values:.2f}")
print(f"Italy bounds: {italy_bounds}")
print(f"Climate data coordinates: {finland_data.longitude.min().values:.2f}  {finland_data.latitude.min().values:.2f} {finland_data.longitude.max().values:.2f}  {finland_data.latitude.max().values:.2f}")
print(f"Finland bounds: {finland_bounds}")


### Basic data visualization with Xarray and Matplotlib

Let's start with a simple visualization: the mean temperature In Italy and Finland averaged over the entire time period (1950-2024).

In [ ]:
# Create a multi-panel plot
fig, axes = plt.subplots(1, 2, figsize=(18, 6))

# Italy mean temperature map
italy_mean = italy_data.tg.mean(dim='time')
im1 = italy_mean.plot(ax=axes[0], cmap='RdBu_r', 
                     vmin=-5, vmax=20,
                     cbar_kwargs={'label': '°C'})
axes[0].set_title('Mean Temperature in Italy (°C)')
italy.boundary.plot(ax=axes[0], color='black', linewidth=1)

# Finland mean temperature map
finland_mean = finland_data.tg.mean(dim='time')
im2 = finland_mean.plot(ax=axes[1], cmap='RdBu_r', 
                       vmin=-5, vmax=20,
                       cbar_kwargs={'label': '°C'})
axes[1].set_title('Mean Temperature in Finland (°C)')
finland.boundary.plot(ax=axes[1], color='black', linewidth=1)

plt.show()

### Time series analysis of temperature trends

Now let's examine annual mean temperature trends over time for both countries.

In [ ]:
# Calculate spatial means for each time step
# Creates a 1D time series where each value is the spatial mean temperature
italy_timeseries = italy_data.tg.mean(dim=['latitude', 'longitude'])
finland_timeseries = finland_data.tg.mean(dim=['latitude', 'longitude'])

# Calculate annual means: uses xarray's resample with 'YE' (Year End) frequency to group data by year and take the mean
italy_annual = italy_timeseries.resample(time='YE').mean()
finland_annual = finland_timeseries.resample(time='YE').mean()

# Plot the time series
plt.figure(figsize=(14, 6))
italy_annual.plot(label='Italy', linewidth=2, color="MediumBlue")
finland_annual.plot(label='Finland', linewidth=2, color="green")
plt.grid(True, linestyle='--', alpha=0.7)
plt.title('Annual Mean Temperature (1950-2024)', fontsize=16)
plt.ylabel('Temperature (°C)', fontsize=12)
plt.xlabel('Year', fontsize=12)
plt.legend(fontsize=12)
plt.show()

## Precipitation levels and drought conditions

**Droughts** represent one of the most impactful natural disasters. They can cause serious problems for the economy, the environment, and people’s lives. As climate change continues to affect weather patterns and rainfall, many regions are facing droughts that are becoming more frequent and more intense.

Drought **indices** are tools to measure how severe a drought is. They help us understand the current state of droughts by evaluating various climate conditions. 
Theo indicate the severity, location, timing, and duration of droughts

### SPEI index

The **Standardized Precipitation Evapotranspiration Index** (**SPEI**) is a widely used meteorological index for measuring drought conditions. The **SPEI** scores indicate how wet or dry a period is compared to normal conditions: negative values indicate drier than usual periods while positive values correspond to wetter than usual periods.

Our main focus will be on understanding how droughts have varied both geographically and temporally.

#### Import the libraries

We need to import further libraries in addition to those used in step 1.

#### Get data from XDS

The ERA5-Drought data is provided by the [Cross Data Store (XDS)](https://xds.ecmwf.int/datasets/derived-drought-historical-monthly?tab=overview), a free online platform provided by the European Centre for Medium-Range Weather Forecasts (ECMWF).

All data and tools available on the XDS are free to use, but an account is required to access and interact with the platform. 

Here is the [link to the registration page](https://www.ecmwf.int/user/login)

In [ ]:
# Standard library imports
import glob
import math
import zipfile
import warnings


# Data manipulation
import dask


# Geospatial libraries
import cartopy.crs as ccrs
import cartopy.feature as cfeature


# Plotting libraries
import plotly.graph_objects as go

# IPython utilities
import ipywidgets as widgets
from IPython.display import display, Markdown

# Suppress specific warnings
warnings.filterwarnings("ignore", category=RuntimeWarning)

In [ ]:
def download_cds_data(area: list[float], year_range: list[int]) -> None:
    """
    Downloads the SPEI (Standardised Precipitation Evapotranspiration Index) 
    for the month of August from 2000 to 2024 for a specified geographical area 
    using the Copernicus Climate Data Store API.

    Parameters:
    - area (List[float]): Bounding box of the area in the format 
      [max_latitude, min_longitude, min_latitude, max_longitude].

    Returns:
    - None
    """
    URL = 'https://xds-preprod.ecmwf.int/api'  # Pre-operational API endpoint for accessing data. Dataset DOI: 10.24381/9bea5e16
    KEY = '<YOUR_KEY>'

    cdsapi_kwargs = {
        'url': URL,
        'key': KEY,
    }
    dataset = "derived-drought-historical-monthly"
    request = {
        "variable": ["standardised_precipitation_evapotranspiration_index"],
        "accumulation_period": ["12"],
        "version": "1_0",
        "product_type": ["reanalysis"],
        "dataset_type": "consolidated_dataset",
        "year": [str(y) for y in range(year_range[0], year_range[-1])],
        "month": ["08"],
        "area": area  # [max_lat, min_lon, min_lat, max_lon]
    }

    client = cdsapi.Client(**cdsapi_kwargs)
    client.retrieve(dataset, request).download()

In [ ]:
year_range = [2000, 2025]
area_coords = [70.0, 19.5, 59.75, 31.5]  # [max_lat, min_lon, min_lat, max_lon]
download_cds_data(area_coords, year_range)

In [ ]:
def extract_zip(unique_code: str) -> None:
    """
    Extracts the contents of a ZIP file into a directory named using the pattern 'data-{unique_code}'.

    The directory name is constructed by removing the '.zip' extension from the given filename.
    If the directory does not exist, it is created. The contents of the ZIP file are extracted there.

    Parameters:
    - zip_filename (str): Name of the ZIP file to extract.

    Returns:
    - None
    """
    data_dir = f"data-{unique_code}"

    if not os.path.exists(data_dir):
        os.makedirs(data_dir)

    zip_filename = f"{unique_code}.zip"
    with zipfile.ZipFile(zip_filename, 'r') as zip_ref:
        zip_ref.extractall(data_dir)

In [ ]:
unique_data_code = "83d599de5c73a5db89f633fe5fee880c"
extract_zip(unique_data_code)

In [ ]:
def merge_netcdf_files(data_dir, file_pattern, parallel=False):
    """
    Merge multiple netCDF files using xarray.
    
    Parameters:
    data_dir : str
        Directory containing the netCDF files
    file_pattern : str
        Pattern to match the netCDF files
    parallel : bool, default=True
        Whether to use parallel processing for opening files
        
    Returns:
    xr.Dataset
        Merged dataset containing all the data

    Note:
    Setting parallel=True in xarray.open_mfdataset() can cause kernel crashes due to high memory usage, especially when working with many or large netCDF files.  If crashes occur repeatedly, set parallel=False.
    """
    file_pattern_path = os.path.join(data_dir, file_pattern)
    files = sorted(glob.glob(file_pattern_path))   # Sorting by time to ensure chronological order
    
    print(f"Found {len(files)} files to merge")
    
    # Open and merge all files in a memory-efficient way using open_mfdataset
    # The 'combine' parameter is set to 'nested' for time series data
    # 'concat_dim' parameter will automatically detect the dimension to concatenate along
    ds = xr.open_mfdataset(
        files,
        parallel=parallel,
        combine='nested',  
        concat_dim='time',  
        engine='netcdf4',
        chunks={'time': 'auto'}  # Use dask for chunking along time dimension
    )
    
    # Ensure the data is properly sorted by time
    if 'time' in ds.dims:
        ds = ds.sortby('time')
    
    return ds

In [ ]:
file_pattern = "SPEI12_genlogistic_global_era5_moda_ref1991to2020_*.nc"
ds = merge_netcdf_files(f"data-{unique_data_code}", file_pattern)

display(ds)

#### Explore our dataset

In [ ]:
print("Global Attributes:")
ds.attrs

In [ ]:
ds.data_vars

#### Preprocess data (possible troubles you may encounter)

In [ ]:
def replace_invalid_values(data: pd.DataFrame, invalid_value: float = -9999.0) -> (pd.DataFrame, int, float):
    """
    Replaces specified invalid values in a pandas DataFrame with NaN (Not a Number), counts the number of replacements,
    and calculates the ratio of invalid values over the total number of values.

    Parameters:
    - data (pd.DataFrame): The input DataFrame containing possibly invalid values.
    - invalid_value (float, optional): The value to be considered as invalid and replaced. 
      Defaults to -9999.0.

    Returns:
    - pd.DataFrame: A DataFrame with invalid values replaced by NaN.
    - int: The number of invalid values replaced.
    - float: Ratio of invalid values to total values in the DataFrame.
    """
    # Create a mask to identify invalid values
    mask = data == invalid_value
    # Count invalid entries across all columns, using compute() if necessary
    if 'dask' in str(type(data)):  # Check if the DataFrame is a Dask DataFrame
        invalid_count = mask.sum().compute()  # Use Dask's compute() for Dask DataFrames
    else:
        invalid_count = mask.sum().sum()  # Sum all true values in the mask for Pandas DataFrame

    # Calculate total number of values in the DataFrame
    total_values = data.size

    # Calculate the ratio of invalid values
    invalid_ratio = invalid_count / total_values

    # Replace invalid values with NaN using where()
    clean_data = data.where(~mask, np.nan)  # Replace where mask is not True
    
    return clean_data, int(invalid_count), float(invalid_ratio)



def remove_time_duplicates(dataset: xr.DataArray) -> (xr.DataArray, int):
    """
    Removes duplicate time entries from an xarray DataArray by keeping the first occurrence
    of each time point and discarding the subsequent duplicates. Also returns the number of duplicates removed.

    Parameters:
    - dataset (xr.DataArray): The input DataArray that contains a time dimension with potentially duplicated entries.

    Returns:
    - xr.DataArray: A new DataArray with the same data as the input but with duplicate time entries removed.
    - int: The number of duplicate time entries removed.
    
    Note:
    - This function is needed because the value of the variable 'time' of January 2024 is duplicated
    """
    time_series = pd.Series(dataset.time.values)
    original_count = len(time_series)
    unique_indices = time_series.drop_duplicates(keep='first').index
    new_dataset = dataset.isel(time=unique_indices)
    removed_count = original_count - len(unique_indices)
    
    return new_dataset, removed_count



def process_xr_datarray(data_array: xr.DataArray) -> (xr.DataArray, dict):
    """
    Processes an xarray DataArray through a sequence of data cleaning and transformation steps
    to ensure its usability in further analysis or modeling. This function standardizes the DataArray
    and returns the counts of cleaned, removed, and converted values.

    Parameters:
    - data_array (xr.DataArray): The input DataArray that will undergo processing.

    Returns:
    - xr.DataArray: The processed DataArray with standardized data formatting and cleaned values.
    - dict: Dictionary containing counts of cleaned, removed, and converted entries.
    """
    data_array, invalid_replaced_count, invalid_ratio = replace_invalid_values(data_array)
    data_array, duplicate_removed_count = remove_time_duplicates(data_array)

    return data_array, {
        'invalid_values_replaced': invalid_replaced_count,
        'invalid_ratio': round((invalid_ratio * 100), 2),
        'duplicates_removed': duplicate_removed_count,
    }

In [ ]:
index = list(ds.data_vars)[0]
processed_ds, change_summary = process_xr_datarray(ds[index])
display(Markdown('**Change summary:**'))
for key, val in change_summary.items():
    print(key, val)

#### Mapping the geographical distribution over time

An effective way to assess drought conditions is by visually examining a series of maps in a timelapse format, with one map representing each year for the chosen month. This allows you to easily observe changes and trends in drought severity over time.

The following function `plot_geographical_distribution` illustrates drought conditions using two distinct colors: brown for drought and green for wet areas. The intensity of the colors indicates the severity of the conditions: deeper shades of brown represent more severe drought, while deeper shades of green indicate wetter conditions.

The geographic area is divided into regular squares, each corresponding to the 0.25-degree grid of the climate mils).

Run these cells and explore the result with the slider bar a the top. Here some suggestion on what you can look for and examine using these maps:
- **Identify drought patterns**: look for areas where the brown color is consistently deep or widespread, indicating regions that are experiencing severe and prolonged drought conditions.

- **Track changes over time**: use the timelapse feature to observe how drought conditions evolve over the years. Pay attention to any patterns of expansion or contraction in the drought areas.

- **Compare wet and dry periods**: contrast periods where green (wet) areas dominate with those where brown (dry) areas are more prevalent. This can help identify cycles of wet and dry periods.

- **Regional analysis**: focus on specific regions within the geographic area to see how drought conditions vary from one place to another. Some regions may be more prone to drought than others.

- **Seasonal Variations**: if the maps cover different seasons within each year, examine how drought conditions change with the seasons. This can provide insights into the impact of seasonal climate patterns.

- **Long-term Trends**: examine the maps for any long-term trends, such as increasing or decreasing drought severity over decades. This can be valuable for understanding the broader climate trends in the region.

- **Anomalies**: look for any unusual patterns or anomalies in the data, such as sudden changes in drought conditions that might indicate an unusual event or error in data collection.

In [ ]:
def plot_geographical_distribution(data: xr.DataArray) -> None:
    """
    Display an interactive map of the SPEI12 drought index for a geographic area using a year slider.

    This function visualizes the geographical distribution of the Standardized 
    Precipitation-Evapotranspiration Index (SPEI12) for a range of years. 
    It generates a map for each year and allows interactive 
    exploration using a slider widget labeled with actual years. The map uses 
    a diverging 'BrBG' colormap to represent dry (brown) and wet (green) anomalies.

    Parameters:
    data : xarray.DataArray
        A 3D DataArray with dimensions (time, lat, lon) containing SPEI12 values.
        The 'time' coordinate should correspond to August of each year.

    Returns:
    None
        Displays an interactive plot.
    """
    # Extract years from the time dimension
    years = data.time.dt.year.values

    # Define the plotting function for a specific year value
    def plot_for_year(year):
        index = list(years).index(year)
        yearly_data  = data.isel(time=index)

        fig = plt.figure(figsize=(8, 6))
        ax = plt.axes(projection=ccrs.PlateCarree())

        # Plotting
        yearly_data .plot(
            ax=ax,
            transform=ccrs.PlateCarree(),
            cmap='BrBG',
            vmin=-2.5, vmax=2.5,
            cbar_kwargs={'label': 'SPEI12'},
            add_colorbar=True
        )

        # Add map features
        ax.set_title(f"SPEI12 - August {year}", fontsize=14)
        ax.coastlines()
        ax.add_feature(cfeature.BORDERS, linestyle=':')
        ax.add_feature(cfeature.LAND, edgecolor='black', alpha=0.1)
        ax.gridlines(draw_labels=True)

        plt.show()

    # Use actual year labels in the slider
    slider = widgets.SelectionSlider(
        options=list(years),
        value=years[0],
        description='Year:',
        continuous_update=False,
        style={'description_width': 'initial'}
    )

    # Interactive display
    widgets.interact(plot_for_year, year=slider)

In [ ]:
plot_geographical_distribution(processed_ds)

#### Exploring trends with median and mean values

To validate your observations and insights, you can check some key statistical measures. The `compute_stats` function calculates the most significant statistics for your data subset. For now, we'll focus on the mean and the median.

The **mean** (average) gives you the average level of drought severity across the entire geographic area for each year relative to the considered month. This helps you understand the general intensity of drought across the whole region or during a specific period. However, the mean can be influenced by extreme values, so it might not always represent the "typical" condition.

The **median** is the middle value when all data points are sorted from least to most severe. Unlike the mean, the median isn't affected by outliers or extreme values. In the context of drought, the median gives you a better sense of the "typical" condition in the region. For instance, if the median drought index is low, it suggests that more than half of the area is experiencing less severe drought, even if a few parts are much worse.

By comparing the mean and median, you can get a clearer picture of drought conditions. If the mean is much higher than the median, it suggests that there are some very severe droughts in parts of the area that are raising the average, even if most of the region is less affected. Conversely, if the mean and median are close, it indicates a more consistent level of drought severity across the area.

In [ ]:
def compute_stats(data: xr.DataArray, full_stats: bool = True) -> dict:
    """
    Computes the basic statistics from the SPEI data over latitude and longitude.
    These statistics include the median, lower and upper quantiles (25th and 75th percentiles), minimum, and maximum values.

    Parameters:
    data (xr.DataArray): The DataArray containing the SPEI data with dimensions including 'lat', 'lon', and 'time'.
    full_stats (bool): If True, computes all statistics. If False, computes only mean and median.

    Returns:
    dict: A dictionary containing:
        - times (np.ndarray): The array of time values.
        - means (np.ndarray): The array of mean values over the specified dimensions.
        - medians (np.ndarray): The array of median values over the specified dimensions.
        - q1s (np.ndarray): The array of 25th percentile values (only if full_stats is True).
        - q3s (np.ndarray): The array of 75th percentile values (only if full_stats is True).
        - mins (np.ndarray): The array of minimum values (only if full_stats is True).
        - maxs (np.ndarray): The array of maximum values (only if full_stats is True).
    """
    # Chunk the data along latitude and longitude to spped up the calculation
    data = data.chunk({'lat': 'auto', 'lon': 'auto'})
    
    # Remove NaN values across lat and lon dimensions for more robust stats
    valid_data = data.dropna(dim='lat', how='all').dropna(dim='lon', how='all')
    
    # Initialize the result dictionary
    result = {}

    # Compute the median and mean together to avoid recomputation
    median = valid_data.median(dim=['lat', 'lon'], skipna=True)
    mean = valid_data.mean(dim=['lat', 'lon'], skipna=True)

    # Compute additional statistics if full_stats is True
    if full_stats:
        q1 = valid_data.quantile(0.25, dim=['lat', 'lon'], skipna=True)
        q3 = valid_data.quantile(0.75, dim=['lat', 'lon'], skipna=True)
        min_val = valid_data.min(dim=['lat', 'lon'], skipna=True)
        max_val = valid_data.max(dim=['lat', 'lon'], skipna=True)
        
        # Compute all stats at once, parallelized
        median_computed, mean_computed, q1_computed, q3_computed, min_computed, max_computed = dask.compute(
            median, mean, q1, q3, min_val, max_val
        )
        
        result.update({
            'means': mean_computed.values,
            'q1s': q1_computed.values,
            'q3s': q3_computed.values,
            'mins': min_computed.values,
            'maxs': max_computed.values,
        })
    else:
        median_computed, mean_computed = dask.compute(median, mean)

    # Update the result dictionary with median and time values
    result.update({
        'times': median_computed['time'].values if 'time' in median_computed.dims else None,
        'medians': median_computed.values,
    })

    return result

In [ ]:
stat_values = compute_stats(processed_ds)

To visually grasp these concepts, we use a scatterplot chart that displays the mean (circle symbol) and the median (diamond symbol) over the years.

In [ ]:
cmap = {
    "Extremely wet": "#074B30",
    "Severely wet": "#0D965f",
    "Moderately wet": "#43EFAA",
    "Mildly wet": "#A1F7D5",
    "Mildly dry": "#FFBF69",
    "Moderately dry": "#FF961F",
    "Severely dry": "#8F5100",
    "Extremely dry": "#3d2100"
    }


def assign_colors(values):
    """
    Assigns colors based on the input values.
    This function takes a dictionary of values and assigns a color code to each value based on the degree of wetness or dryness.
    If the value is NaN, the color will be transparent (color: 'rgba(0,0,0,0)').

    Parameters:
    cmap (dict of float): A dictionary of colors from the json file.

    Returns:
    list of str: A list of color codes corresponding to the values.
    
    Note: 
    neutral: #B89A7D - axes: #D3D3D3
    """
  
    colors = []
    for val in values:
        if math.isnan(val):
            colors.append('rgba(0,0,0,0)')  # transparent for NaN values
        elif val > 2.0:
            colors.append(cmap["Extremely wet"])
        elif 1.5 < val <= 2.0:
            colors.append(cmap["Severely wet"])
        elif 1.0 < val <= 1.5:
            colors.append(cmap["Moderately wet"])
        elif 0 < val <= 1.0:
            colors.append(cmap["Mildly wet"])
        elif -1.0 < val <= 0:
            colors.append(cmap["Mildly dry"])
        elif -1.5 < val <= -1.0:
            colors.append(cmap["Moderately dry"])
        elif -2.0 < val <= -1.5:
            colors.append(cmap["Severely dry"])
        elif val <= -2.0:
            colors.append(cmap["Extremely dry"])
    return colors


def create_scatterplot(values: dict):
    """
    Creates and displays a scatterplot of the Standardized Precipitation-Evapotranspiration Index (SPEI)
    over time, using provided data points for both mean and median values. The plot marks mean values
    with dots and median values with diamonds, each colored according to their SPEI categories. Tooltips
    provide additional data insights on hover.

    Parameters:
    - values (dict): Contains time series data for the plot, including:
        - 'times': List of datetime objects representing the years.
        - 'means': List of mean SPEI values corresponding to the times.
        - 'medians': List of median SPEI values corresponding to the times.

    Returns:
    - None: This function creates and displays the scatterplot directly using the Plotly visualization library.
    """
    times = values['times']
    means = values['means']
    medians = values['medians']
    
    
    # Assign colors to means and medians based on their SPEI values
    mean_colors = assign_colors(means)
    median_colors = assign_colors(medians)
    
    # Prepare the text for the tooltips
    # Convert numpy.datetime64 for formatting
    mean_tooltip_texts = [f"{np.datetime64(time, 'D').astype('datetime64[M]').astype(object).strftime('%B %Y')}, mean: {mean:.2f}" for time, mean in zip(times, means)]
    median_tooltip_texts = [f"{np.datetime64(time, 'D').astype('datetime64[M]').astype(object).strftime('%B %Y')}, median: {median:.2f}" for time, median in zip(times, medians)]
        
        
    # Define the scatter plot for means
    mean_trace = go.Scatter(
        x=times,
        y=means,
        mode='markers',
        name='Mean SPEI',
        marker=dict(
            size=10,
            symbol='circle',
            color=mean_colors,
            line=dict(width=1, color='black')  # Border
        ),
        text=mean_tooltip_texts,  # Custom tooltip text
        hoverinfo='text'
    )
    
    # Define the scatter plot for medians
    median_trace = go.Scatter(
        x=times,
        y=medians,
        mode='markers',
        name='Median SPEI',
        marker=dict(
            size=10,
            symbol='diamond',
            color=median_colors,
            line=dict(width=1, color='black')  # Border
        ),
        text=median_tooltip_texts,  # Custom tooltip text
        hoverinfo='text'
    )
    
    
    # Initialize and update the figure
    fig = go.Figure([mean_trace, median_trace])
    
    # Add dummy traces for color legend using the spei categories
    for category, color in cmap.items():
        fig.add_trace(go.Scatter(
            x=[None], y=[None], mode='markers',
            marker=dict(size=10, color=color),
            name=category,
            legendgroup='colors', showlegend=True,
            hoverinfo='none'
    ))
    
    # Update the figure
    fig.update_layout(
        title=f"SPEI12, trends over time in the month of August",
        xaxis_title='Years',
        yaxis_title=f"SPEI12 Value",
        height=600,
        template='plotly_white',
        plot_bgcolor='rgba(0,0,0,0)',
        yaxis=dict(
            zeroline=True,
            zerolinewidth=1,
            zerolinecolor='#D3D3D3'
        )
    )

    # Display the figure
    fig.show()

In [ ]:
create_scatterplot(stat_values)

#### Combining years

Finally, to have a more comprehensive view of the drought phenomenon in a specific area, combining data across years (and even months) in a stripe chart. These charts allow us to quickly identify temporal patterns, showing how drought conditions evolve over years or months.

In [ ]:
def create_stripechart(values: dict, year_range: list):
    """
    Generate and display a stripe chart showing temporal trends in median values for a specific climate index.

    This function visualizes monthly median values across multiple years using color-coded bars (stripes),
    where each color represents a category (e.g., based on drought or wetness levels from an index like SPEI).
    It includes dynamic tooltips, an adaptive time axis, and a legend that decodes the color scale.

    Parameters:
        values (dict): A dictionary containing:
            - 'times' (list-like): List of datetime-like objects representing observation times.
            - 'medians' (list-like): List of median index values corresponding to each time.
        year_range (list): A list with two elements: start year and end year (e.g., [2000, 2024]).

    Returns:
        None: The function displays the chart using Plotly but does not return an object.

    Notes:
        - This chart is useful for showing long-term seasonal patterns and anomalies in a single month over years.
        - Tick labels on the x-axis are adapted dynamically based on the length of the year range.
        - The function assumes a predefined color map (`cmap`) and uses a helper function `assign_colors()` 
          to assign categories based on median values.
    """
    times = values['times']
    medians = values['medians']
    colors = assign_colors(medians)
    
    # Convert numpy datetime64 to datetime64[M] for sorting
    date_transform = [np.datetime64(time, 'M') for time in times]

    # Pair date_transform, colors, and medians, then sort by date_transform
    paired_data = sorted(zip(date_transform, colors, medians), key=lambda x: x[0])
    sorted_dates = [date for date, _, _ in paired_data]
    sorted_colors = [color for _, color, _ in paired_data]
    sorted_medians = [median for _, _, median in paired_data]
    
    # Generate tooltip texts
    tooltip_texts = [f"{date.astype(object).strftime('%B %Y')}, median: {median:.2f}" for date, median in zip(sorted_dates, sorted_medians)]
    sorted_dates = [str(date) for date in sorted_dates]  # Convert to string for plotting



    # Filter dates to show only specific years in ten-year intervals
    start_year = int(year_range[0])
    end_year = int(year_range[-1])
    year_diff = end_year - start_year
    if year_diff <= 15:
        interval = (3, 1)
    elif year_diff > 15 and year_diff <= 25:
        interval = (5, 2)
    else:
        interval = (10, 10)

    target_years = [str(year) for year in range(start_year, end_year + 1, interval[0])]
    display_dates = [date[:4] if (date[:4] in target_years and date[5:7] == '01') else '' for date in sorted_dates]


    # Create the stripe chart
    bar_width = 1.0
    fig = go.Figure()
    
    
    fig.add_trace(go.Bar(
        x=sorted_dates,
        y=[1] * len(sorted_dates),
        marker_color=sorted_colors,
        hovertext=tooltip_texts,
        hoverinfo='text',
        orientation='v',
        showlegend=False,  # Disable showing each bar in the legend
        width=bar_width
    ))
    
        # Add dummy bars for color legend
    for category, color in cmap.items():
        fig.add_trace(go.Bar(
            x=[None],
            y=[None],
            marker_color=color,
            name=category,
            showlegend=True,  # Enable legend only for dummy bars
            hoverinfo='none'
    ))
        

    # Update layout
    fig.update_layout(
        title=f"Median SPEI12, trends over time in the month of August from {year_range[0]} to {year_range[-1]}",
        xaxis_title='Years',
        yaxis_title=f"Median SPEI12",
        height=600,
        yaxis=dict(showticklabels=False, showgrid=False, zeroline=False),
        xaxis={'type': 'category', 'tickmode': 'array', 'tickvals': sorted_dates, 'ticktext': display_dates, 'tickangle': 0},
        template='plotly_white',
        plot_bgcolor='rgba(0,0,0,0)',
        showlegend=True,
        legend_title="SPEI Categories"
    )
    fig.update_yaxes(visible=False)
    fig.show()   

In [ ]:
create_stripechart(stat_values, year_range)

## 👉 Now it's your turn
Replicate the analysis for Italy. Use the following bounding box: (max_lat=47.0, min_lon=6.75, min_lat=35.5, max_lon=18.5).

In [ ]:
# Download CDS data
year_range = []  # [2000, 2025]
area_coords = [] # Italy (max_lat, min_lon, min_lat, max_lon):  [47.0, 6.75, 35.5, 18.5]
download_cds_data(area_coords, year_range)

In [ ]:
# Extract files from zip
unique_data_code = ""   # 26bb58ba892703300d478eee6fb4a31b
extract_zip(unique_data_code)

In [ ]:
# Merge netCDF files
file_pattern = ""    # e.g. "SPEI12_genlogistic_global_era5_moda_ref1991to2020_*.nc"
ds = merge_netcdf_files(f"data-{unique_data_code}", file_pattern)

display(ds)

In [ ]:
# Preprocess data
index = list(ds.data_vars)[0]
processed_ds, change_summary = process_xr_datarray(ds[index])
display(Markdown('**Change summary:**'))
for key, val in change_summary.items():
    print(key, val)

In [ ]:
plot_geographical_distribution(processed_ds)

In [ ]:
# Compute statistics
stat_values = compute_stats(processed_ds)

In [ ]:
create_scatterplot(stat_values)

In [ ]:
create_stripechart(stat_values, year_range)